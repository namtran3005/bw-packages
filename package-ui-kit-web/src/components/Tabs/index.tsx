export { Tabs, TabsProps, TabsItem, TabsItemProps } from './Tabs';
