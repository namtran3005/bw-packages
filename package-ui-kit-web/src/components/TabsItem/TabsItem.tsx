import * as React from 'react';

interface TabsItemProps {
  children?: any; // eslint-disable-line @typescript-eslint/no-explicit-any
  label: React.ReactChild;
  disabled?: boolean;
}

const TabsItem: React.FunctionComponent<TabsItemProps> = ({
  /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
  label,
  /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
  disabled,
  ...cleanProps
}) => <div {...cleanProps} />;

export { TabsItem, TabsItemProps };
