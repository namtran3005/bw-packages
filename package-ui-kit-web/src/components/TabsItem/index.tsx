export { TabsItem, TabsItemProps } from './TabsItem';
