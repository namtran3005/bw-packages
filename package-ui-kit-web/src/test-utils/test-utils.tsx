import {
  render as rtlRender,
  waitFor,
  RenderOptions as RTLRenderOptions,
} from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import { WebAppThemeProvider } from '../theme';
import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';

type RenderOptions = RTLRenderOptions;

function renderWithTestHarness(
  ui: React.ReactElement,
  renderOptions: RenderOptions = {}
) {
  function Wrapper({ children }: any) {
    return (
      <MemoryRouter>
        <WebAppThemeProvider>{children}</WebAppThemeProvider>
      </MemoryRouter>
    );
  }

  return rtlRender(ui, {
    wrapper: Wrapper,
    ...renderOptions,
  });
}

// Use if you're in a hurry (better than nothing)
// be sure to verify no memory leak from implicit
// waitFor by adding these flags to jest: --runInBand --detectOpenHandles
async function renderAct(ui: React.ReactElement, options: RenderOptions = {}) {
  const view = renderWithTestHarness(ui, options);
  return await waitFor(() => view, {
    timeout: 5000,
    onTimeout: (e) => {
      // eslint-disable-next-line no-console
      console.error(e);
      return e;
    },
  });
}

export * from '@testing-library/react';

export { renderWithTestHarness, renderAct, userEvent };
