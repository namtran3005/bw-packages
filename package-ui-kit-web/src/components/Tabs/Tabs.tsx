import * as React from 'react';
import styled from 'styled-components';
import SwipeableViews from 'react-swipeable-views';
import {
  Colors,
  ColorsType,
  Spacing,
  Transitions,
} from '@bitwala-cryptobank-squad/package-theme';
import { Tabs as MuiTabs, Tab as MuiTab } from '@material-ui/core';
import { TabsProps as MuiTabsProps } from '@material-ui/core/Tabs';

import { Content, Typography } from '../..';
import { TabsItem, TabsItemProps } from '../TabsItem';
import { SpacingType } from '../Content';

interface TabsProps extends RootProps {
  value: number;
  foreground?: ColorsType;
  labelBarStartContent?: React.ReactNode;
  labelBarEndContent?: React.ReactNode;
}

interface RootProps extends Omit<MuiTabsProps, 'onChange' | 'ref'> {
  innerRef?: any; // eslint-disable-line @typescript-eslint/no-explicit-any
  onChange?: (
    event: React.ChangeEvent<{}> | React.FormEvent<HTMLButtonElement>,
    value: number
  ) => void;
  spacing?: SpacingType;
  gap?: SpacingType;
  indicatorForeground?: ColorsType;
}

interface RootLabelBarContentProps {
  fullWidth?: boolean;
  textColor?: string;
  indicator?: React.ReactNode;
}

const defaultSpacing = 24;

const RootLabelBarContentComponent: React.FunctionComponent<RootLabelBarContentProps> = ({
  /* eslint-disable @typescript-eslint/no-unused-vars */
  fullWidth,
  indicator,
  textColor,
  /* eslint-enable @typescript-eslint/no-unused-vars */
  ...cleanProps
}) => <div {...cleanProps} />;

const RootLabelBarContent = styled(RootLabelBarContentComponent)``;

const RootTab = styled(MuiTab).withConfig({
  shouldForwardProp: (prop) => !['indicatorForeground'].includes(prop),
})`
  &&& {
    position: relative;
    text-align: left;
    text-transform: none;
    opacity: 0.5;
    transition: ${Transitions.duration.normal}ms ${Transitions.easing.swiftOut};
    max-width: none;
    min-height: 0;
    height: fit-content;
    padding-left: 0;
    padding-top: 0;
    padding-bottom: ${Spacing.unit / 2};

    &:before {
      content: '';
      position: absolute;
      bottom: 0;
      left: 0;
      opacity: 0;
      transition: ${Transitions.duration.normal}ms
        ${Transitions.easing.swiftOut};
      width: ${Spacing.unit * 2}px;
      height: ${Spacing.unit / 2}px;
    }

    &:focus {
      &:before {
        opacity: 0.4;
      }
    }

    &.selected {
      opacity: 1;

      &:focus {
        &:before {
          opacity: 0;
        }
      }
    }

    &.disabled {
      opacity: 0.1;
    }

    .wrapper {
      align-items: start;
    }
  }
`;

const Root: React.FC<RootProps> = styled(MuiTabs).withConfig({
  shouldForwardProp: (prop) => !['indicatorForeground'].includes(prop),
})`
  &&& {
    margin-bottom: -${({ spacing }) => (typeof spacing !== 'undefined' ? spacing : defaultSpacing)}px;
    padding: ${({ spacing }) =>
      typeof spacing !== 'undefined' ? spacing : defaultSpacing}px;
    min-height: 0;

    .indicator {
      background: ${({ indicatorForeground }: RootProps) =>
        Colors.get(indicatorForeground || 'primaryLilac')};
      width: ${Spacing.unit * 2}px !important;
      height: ${Spacing.unit / 2}px;
    }

    .react-swipeable-view-container > div {
      overflow: visible !important;
    }

    ${RootTab} {
      margin-left: ${({ gap }) => (gap || 0) / 2}px;
      margin-right: ${({ gap }) => (gap || 0) / 2}px;

      &:first-of-type {
        margin-left: 0;
      }

      &:last-of-type {
        margin-right: 0;
      }
    }

    ${RootTab}:focus:before {
      background: ${({ indicatorForeground }) =>
        Colors.get(indicatorForeground || 'primaryLilac')};
    }
  }
`;

const RootTabContainer = Content;

Root.displayName = 'Tabs';
RootTab.displayName = 'TabsTab';
RootTabContainer.displayName = 'TabsTabContainer';

class Tabs extends React.Component<TabsProps, {}> {
  public static Item = TabsItem;

  public renderTabs(): ({} | null | undefined)[] | undefined | null {
    return React.Children.map(this.props.children, (child) => {
      if (!child || !React.isValidElement(child)) {
        return child;
      }
      /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
      const { label, children, ...cleanProps } = child.props as TabsItemProps;
      const { foreground } = this.props;

      return (
        <RootTab
          classes={{
            root: 'root',
            labelIcon: 'labelIcon',
            textColorInherit: 'textColorInherit',
            textColorPrimary: 'textColorPrimary',
            textColorSecondary: 'textColorSecondary',
            selected: 'selected',
            disabled: 'disabled',
            fullWidth: 'fullWidth',
            wrapper: 'wrapper',
          }}
          disableRipple
          label={
            typeof label === 'string' ? (
              <Typography.Header4 foreground={foreground || 'primaryLilac'}>
                {label}
              </Typography.Header4>
            ) : (
              label
            )
          }
          {...cleanProps}
        />
      );
    });
  }

  public renderTabsContainers(): ({} | null | undefined)[] | undefined | null {
    return React.Children.map(this.props.children, (child) => {
      if (!React.isValidElement(child)) {
        return child;
      }

      return <RootTabContainer spacing={this.props.spacing} {...child.props} />;
    });
  }

  public render(): JSX.Element {
    const {
      labelBarStartContent,
      labelBarEndContent,
      ...cleanProps
    } = this.props;

    return (
      <React.Fragment>
        <Root
          onChange={cleanProps.onChange}
          value={cleanProps.value}
          variant="fullWidth"
          classes={{
            root: 'root',
            flexContainer: 'flexContainer',
            scroller: 'scroller',
            fixed: 'fixed',
            scrollable: 'scrollable',
            centered: 'centered',
            scrollButtons: 'scrollButtons',
            scrollButtonsDesktop: 'scrollButtonsDesktop',
            indicator: 'indicator',
          }}
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          {...(cleanProps as any)}
        >
          {labelBarStartContent && (
            <RootLabelBarContent>{labelBarStartContent}</RootLabelBarContent>
          )}

          {this.renderTabs()}

          {labelBarEndContent && (
            <RootLabelBarContent>{labelBarEndContent}</RootLabelBarContent>
          )}
        </Root>
        <SwipeableViews index={cleanProps.value} animateHeight={false} disabled>
          {this.renderTabsContainers()}
        </SwipeableViews>
      </React.Fragment>
    );
  }
}

export { Tabs, TabsProps, TabsItem, TabsItemProps };
