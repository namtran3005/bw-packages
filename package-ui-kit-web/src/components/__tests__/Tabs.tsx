import * as React from 'react';

import { renderWithTestHarness } from '../../test-utils/test-utils';
import { Tabs } from '../Tabs';
import { TabsItem } from '../TabsItem';

const defaultProps = {
  value: 0,
};

describe('Tabs', () => {
  it('should render Tabs', () => {
    const { baseElement } = renderWithTestHarness(
      <Tabs {...defaultProps}>
        <TabsItem label="First Item">First Item Content</TabsItem>
      </Tabs>
    );
    expect(baseElement).toMatchSnapshot();
  });
  it('should render Tabs with only text content', () => {
    const { baseElement } = renderWithTestHarness(
      <Tabs {...defaultProps}>Random Text Content</Tabs>
    );
    expect(baseElement).toMatchSnapshot();
  });
  it('should render Tabs and click with listener', () => {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const onChangeListener = jest.fn();

    const { baseElement } = renderWithTestHarness(
      <Tabs {...defaultProps} onChange={onChangeListener}>
        <TabsItem label="First Tab">Content</TabsItem>
      </Tabs>
    );
    expect(baseElement).toMatchSnapshot();
  });
});
