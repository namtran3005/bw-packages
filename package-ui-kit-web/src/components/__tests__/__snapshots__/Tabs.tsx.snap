// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs should render Tabs 1`] = `
.c3 {
  text-align: left;
}

.c3.c3.c3 {
  font-weight: 600;
  font-size: 1rem;
  color: #BEAAFF;
  font-family: Inter;
  font-style: normal;
  white-space: inherit;
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
  word-wrap: break-word;
}

.c3.c3.c3:hover {
  background: initial;
}

.c3.c3.c3:active {
  background: initial;
}

.c4 {
  position: relative;
  margin-left: 0px;
  max-width: none;
  min-width: none;
  box-sizing: border-box;
  height: auto;
  min-height: auto;
  z-index: initial;
  -webkit-transition: 300ms cubic-bezier(.62,.28,.23,.99);
  transition: 300ms cubic-bezier(.62,.28,.23,.99);
}

.c4 > .{
  box-sizing: border-box;
  margin-top: auto;
  margin-bottom: auto;
  margin-left: inherit;
  margin-right: inherit;
  max-width: auto;
  text-align: inherit;
  height: auto;
  -webkit-transition: 300ms cubic-bezier(.62,.28,.23,.99);
  transition: 300ms cubic-bezier(.62,.28,.23,.99);
  padding-top: 24px;
  padding-bottom: 24px;
  padding-left: 24px;
  padding-right: 24px;
}

.c2.c2.c2 {
  position: relative;
  text-align: left;
  text-transform: none;
  opacity: 0.5;
  -webkit-transition: 300ms cubic-bezier(.62,.28,.23,.99);
  transition: 300ms cubic-bezier(.62,.28,.23,.99);
  max-width: none;
  min-height: 0;
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  padding-left: 0;
  padding-top: 0;
  padding-bottom: 4;
}

.c2.c2.c2:before {
  content: '';
  position: absolute;
  bottom: 0;
  left: 0;
  opacity: 0;
  -webkit-transition: 300ms cubic-bezier(.62,.28,.23,.99);
  transition: 300ms cubic-bezier(.62,.28,.23,.99);
  width: 16px;
  height: 4px;
}

.c2.c2.c2:focus:before {
  opacity: 0.4;
}

.c2.c2.c2.selected {
  opacity: 1;
}

.c2.c2.c2.selected:focus:before {
  opacity: 0;
}

.c2.c2.c2.disabled {
  opacity: 0.1;
}

.c2.c2.c2 .wrapper {
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
}

.c0.c0.c0 {
  margin-bottom: -24px;
  padding: 24px;
  min-height: 0;
}

.c0.c0.c0 .indicator {
  background: #BEAAFF;
  width: 16px !important;
  height: 4px;
}

.c0.c0.c0 .react-swipeable-view-container > div {
  overflow: visible !important;
}

.c0.c0.c0 .c1 {
  margin-left: 0px;
  margin-right: 0px;
}

.c0.c0.c0 .c1:first-of-type {
  margin-left: 0;
}

.c0.c0.c0 .c1:last-of-type {
  margin-right: 0;
}

.c0.c0.c0 .c1:focus:before {
  background: #BEAAFF;
}

<body>
  <div>
    <div
      class="MuiTabs-root root c0"
    >
      <div
        class="MuiTabs-scroller scroller MuiTabs-fixed fixed"
        style="overflow: hidden;"
      >
        <div
          class="MuiTabs-flexContainer flexContainer"
          role="tablist"
        >
          <button
            aria-selected="true"
            class="MuiButtonBase-root MuiTab-root root MuiTab-textColorInherit textColorInherit c1 c2 Mui-selected selected MuiTab-fullWidth fullWidth"
            role="tab"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiTab-wrapper wrapper"
            >
              <h4
                class="MuiTypography-root c3 MuiTypography-h4"
              >
                First Item
              </h4>
            </span>
          </button>
        </div>
        <span
          class="PrivateTabIndicator-root-1 PrivateTabIndicator-colorSecondary-3 MuiTabs-indicator indicator"
          style="left: 0px; width: 0px;"
        />
      </div>
    </div>
    <div
      style="overflow-x: hidden;"
    >
      <div
        class="react-swipeable-view-container"
        style="flex-direction: row; transition: all 0s ease 0s; direction: ltr; display: flex; will-change: transform;"
      >
        <div
          aria-hidden="false"
          data-swipeable="true"
          style="width: 100%; flex-shrink: 0; overflow: auto;"
        >
          <div
            class="c4"
            label="First Item"
          >
            <div
              class=""
            >
              First Item Content
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Tabs should render Tabs and click with listener 1`] = `
<body>
  .c3 {
  text-align: left;
}

.c3.c3.c3 {
  font-weight: 600;
  font-size: 1rem;
  color: #BEAAFF;
  font-family: Inter;
  font-style: normal;
  white-space: inherit;
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
  word-wrap: break-word;
}

.c3.c3.c3:hover {
  background: initial;
}

.c3.c3.c3:active {
  background: initial;
}

.c4 {
  position: relative;
  margin-left: 0px;
  max-width: none;
  min-width: none;
  box-sizing: border-box;
  height: auto;
  min-height: auto;
  z-index: initial;
  -webkit-transition: 300ms cubic-bezier(.62,.28,.23,.99);
  transition: 300ms cubic-bezier(.62,.28,.23,.99);
}

.c4 > .{
  box-sizing: border-box;
  margin-top: auto;
  margin-bottom: auto;
  margin-left: inherit;
  margin-right: inherit;
  max-width: auto;
  text-align: inherit;
  height: auto;
  -webkit-transition: 300ms cubic-bezier(.62,.28,.23,.99);
  transition: 300ms cubic-bezier(.62,.28,.23,.99);
  padding-top: 24px;
  padding-bottom: 24px;
  padding-left: 24px;
  padding-right: 24px;
}

.c2.c2.c2 {
  position: relative;
  text-align: left;
  text-transform: none;
  opacity: 0.5;
  -webkit-transition: 300ms cubic-bezier(.62,.28,.23,.99);
  transition: 300ms cubic-bezier(.62,.28,.23,.99);
  max-width: none;
  min-height: 0;
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  padding-left: 0;
  padding-top: 0;
  padding-bottom: 4;
}

.c2.c2.c2:before {
  content: '';
  position: absolute;
  bottom: 0;
  left: 0;
  opacity: 0;
  -webkit-transition: 300ms cubic-bezier(.62,.28,.23,.99);
  transition: 300ms cubic-bezier(.62,.28,.23,.99);
  width: 16px;
  height: 4px;
}

.c2.c2.c2:focus:before {
  opacity: 0.4;
}

.c2.c2.c2.selected {
  opacity: 1;
}

.c2.c2.c2.selected:focus:before {
  opacity: 0;
}

.c2.c2.c2.disabled {
  opacity: 0.1;
}

.c2.c2.c2 .wrapper {
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
}

.c0.c0.c0 {
  margin-bottom: -24px;
  padding: 24px;
  min-height: 0;
}

.c0.c0.c0 .indicator {
  background: #BEAAFF;
  width: 16px !important;
  height: 4px;
}

.c0.c0.c0 .react-swipeable-view-container > div {
  overflow: visible !important;
}

.c0.c0.c0 .c1 {
  margin-left: 0px;
  margin-right: 0px;
}

.c0.c0.c0 .c1:first-of-type {
  margin-left: 0;
}

.c0.c0.c0 .c1:last-of-type {
  margin-right: 0;
}

.c0.c0.c0 .c1:focus:before {
  background: #BEAAFF;
}

<div>
    <div
      class="MuiTabs-root root c0"
    >
      <div
        class="MuiTabs-scroller scroller MuiTabs-fixed fixed"
        style="overflow: hidden;"
      >
        <div
          class="MuiTabs-flexContainer flexContainer"
          role="tablist"
        >
          <button
            aria-selected="true"
            class="MuiButtonBase-root MuiTab-root root MuiTab-textColorInherit textColorInherit c1 c2 Mui-selected selected MuiTab-fullWidth fullWidth"
            role="tab"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiTab-wrapper wrapper"
            >
              <h4
                class="MuiTypography-root c3 MuiTypography-h4"
              >
                First Tab
              </h4>
            </span>
          </button>
        </div>
        <span
          class="PrivateTabIndicator-root-9 PrivateTabIndicator-colorSecondary-11 MuiTabs-indicator indicator"
          style="left: 0px; width: 0px;"
        />
      </div>
    </div>
    <div
      style="overflow-x: hidden;"
    >
      <div
        class="react-swipeable-view-container"
        style="flex-direction: row; transition: all 0s ease 0s; direction: ltr; display: flex; will-change: transform;"
      >
        <div
          aria-hidden="false"
          data-swipeable="true"
          style="width: 100%; flex-shrink: 0; overflow: auto;"
        >
          <div
            class="c4"
            label="First Tab"
          >
            <div
              class=""
            >
              Content
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Tabs should render Tabs with only text content 1`] = `
<body>
  .c0.c0.c0 {
  margin-bottom: -24px;
  padding: 24px;
  min-height: 0;
}

.c0.c0.c0 .indicator {
  background: #BEAAFF;
  width: 16px !important;
  height: 4px;
}

.c0.c0.c0 .react-swipeable-view-container > div {
  overflow: visible !important;
}

.c0.c0.c0 .sc-bUrJUP {
  margin-left: 0px;
  margin-right: 0px;
}

.c0.c0.c0 .sc-bUrJUP:first-of-type {
  margin-left: 0;
}

.c0.c0.c0 .sc-bUrJUP:last-of-type {
  margin-right: 0;
}

.c0.c0.c0 .sc-bUrJUP:focus:before {
  background: #BEAAFF;
}

<div>
    <div
      class="MuiTabs-root root c0"
    >
      <div
        class="MuiTabs-scroller scroller MuiTabs-fixed fixed"
        style="overflow: hidden;"
      >
        <div
          class="MuiTabs-flexContainer flexContainer"
          role="tablist"
        />
        <span
          class="PrivateTabIndicator-root-5 PrivateTabIndicator-colorSecondary-7 MuiTabs-indicator indicator"
          style="left: 0px; width: 0px;"
        />
      </div>
    </div>
    <div
      style="overflow-x: hidden;"
    >
      <div
        class="react-swipeable-view-container"
        style="flex-direction: row; transition: all 0s ease 0s; direction: ltr; display: flex; will-change: transform;"
      >
        <div
          aria-hidden="false"
          data-swipeable="true"
          style="width: 100%; flex-shrink: 0; overflow: auto;"
        >
          Random Text Content
        </div>
      </div>
    </div>
  </div>
</body>
`;
